### PR TITLE
TOPdesk integration

### PR DESCRIPTION
**Pull request by DinaMeylakh**
WIP: Does not include mirroring yet.

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/22902

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+
+''' FETCH & MIRRORING COMMANDS'''
+
+
+def fetch_incidents(client: Client,
+                    last_run: Dict[str, Any],
+                    demisto_params: Dict[str, Any]) -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
+    """Fetches incidents from TOPdesk."""
+
+    first_fetch_datetime = dateparser.parse(demisto_params.get('first_fetch', '3 days'))
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:697]: you should use `parse_date_range` from common server python it basically does this for you.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+                                                      max_fetch=int(demisto_params.get('max_fetch', 10)),
+                                                      query=demisto_params.get('query', None),
+                                                      modification_date_start=demisto_params.get(
+                                                          'modification_date_start', None),
+                                                      modification_date_end=demisto_params.get(
+                                                          'modification_date_end', None))
+
+    for topdesk_incident in topdesk_incidents:
+        if topdesk_incident.get('creationDate', None):
+            creation_datetime = dateparser.parse(topdesk_incident['creationDate'])
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:721]: I think we checked but just to make sure - does this come in UTC? or is this connected to a timezone?

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 3]
```diff
@@ -0,0 +1,862 @@
...
+                if isinstance(topdesk_incident_value, str):
+                    labels.append({
+                        'type': topdesk_incident_field,
+                        'value': topdesk_incident_value
+                    })
+                else:
+                    labels.append({
+                        'type': topdesk_incident_field,
+                        'value': json.dumps(topdesk_incident_value)
+                    })
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:737]: If I get where you are aiming, this is used to populate the incident fields right?
if so then there is really no need for this, the incoming mapper should do the trick by taking the rawJSON and using it to map to the fields.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 4]
```diff
@@ -0,0 +1,862 @@
...
+                'name': f"TOPdesk incident {topdesk_incident['number']}",
+                'details': json.dumps(topdesk_incident),
+                'occurred': incident_created_time.strftime(DATE_FORMAT),
+                'rawJSON': json.dumps(topdesk_incident),
+                'labels': labels
+            })
+        if latest_created_time.timestamp() < incident_created_time.timestamp():
+            latest_created_time = incident_created_time
+
+    return {'last_fetch': latest_created_time.strftime(DATE_FORMAT)}, incidents
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:749]: by moving to str here you are losing millisecond info and this can result in duplicated incidents - do we know that topdesk does not use millisec? if not then you should keep the data in a different format like the datetime object or numeric timestamp.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 5]
```diff
@@ -0,0 +1,862 @@
...
+            raise ValueError(f"Cannot get list of type {list_type}.\n "
+                             f"Only {allowed_list_type} are allowed.")
+
+        url_suffix = f"/{list_type}"
+        inline_parameters = False
+        request_params: Dict[str, Any] = {}
+        if start:
+            url_suffix = f"{url_suffix}?start={start}"
+            inline_parameters = True
+        if page_size:
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:54]:
```suggestion

        if page_size:
```
drop a line between if blocks - this will make your code much more readable and easier to actually look at.
This true for all if blocks.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 6]
```diff
@@ -0,0 +1,5907 @@
...
+display: TOPdesk
+name: TOPdesk
+script:
+  commands:
+  - arguments: []
+    deprecated: false
+    description: Get list of subcategories.
+    name: topdesk-subcategories-list
+    outputs:
+    - contextPath: TOPdesk.subcategory..id
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:76]:
```suggestion
    - contextPath: TOPdesk.subcategory.id
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 7]
```diff
@@ -0,0 +1,5907 @@
...
+      description: ''
+      type: String
+    - contextPath: TOPdesk.subcategory..name
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:79]:
```suggestion
    - contextPath: TOPdesk.subcategory.name
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 8]
```diff
@@ -0,0 +1,5907 @@
...
+      description: ''
+      type: String
+    - contextPath: TOPdesk.subcategory..category.id
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:82]:
```suggestion
    - contextPath: TOPdesk.subcategory.category.id
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 9]
```diff
@@ -0,0 +1,5907 @@
...
+      description: ''
+      type: String
+    - contextPath: TOPdesk.subcategory..category.name
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:85]:
```suggestion
    - contextPath: TOPdesk.subcategory.category.name
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 10]
```diff
@@ -0,0 +1,5907 @@
...
+name: TOPdesk
+script:
+  commands:
+  - arguments: []
+    deprecated: false
+    description: Get list of subcategories.
+    name: topdesk-subcategories-list
+    outputs:
+    - contextPath: TOPdesk.subcategory..id
+      description: ''
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:77]: you are missing descriptions for all context outputs. please fill them.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 11]
```diff
@@ -0,0 +1,5907 @@
...
+      type: String
+    - contextPath: TOPdesk.subcategory..category.name
+      description: ''
+      type: String
+  - arguments: []
+    deprecated: false
+    description: Get list of categories.
+    name: topdesk-categories-list
+    outputs:
+    - contextPath: TOPdesk.category..id
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:93]:
```suggestion
    - contextPath: TOPdesk.category.id
```
not sure how you got this double dot situation -but please make sure you are using only one dot between context levels.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 12]
```diff
@@ -0,0 +1,5907 @@
...
+  dockerimage: demisto/python3:3.9.1.15759
+  isfetch: true
+  longRunning: false
+  longRunningPort: false
+  runonce: false
+  script: '-'
+  subtype: python3
+  type: python
+defaultclassifier: HelloWorld
+defaultmapperin: HelloWorld-mapper
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:5904]: change you your classifier and mapper.

**Review by roysagi — CHANGES_REQUESTED** [review]
All in all a good first draft here are some keynotes:

1) Please Capitalize the context keys being used it will make the output look better, same is true for the readable output - you can actually use the same dict and just put the fields you want to display in the headers for the readable output.

2) Please add raw_response for all the command outputs.

3) Drop a line after if blocks - this will make your code more readable.

4) You are missing descriptions for all your context outputs and also you have two dots separating context level instead of one.

When you get back we should sit to review my notes here.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 13]
```diff
@@ -0,0 +1,862 @@
...
+            else:
+                url_suffix = f"{url_suffix}?page_size={page_size}"
+                inline_parameters = True
+        if query:
+            if list_type != "incidents" or self.new_query:
+                query = f"query={query}"
+            if inline_parameters:
+                url_suffix = f"{url_suffix}&{query}"
+            else:
+                url_suffix = f"{url_suffix}?{query}"
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:66]: start, page_size and query are inline unfortunately

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 14]
```diff
@@ -0,0 +1,862 @@
...
+            raise e
+
+
+def incident_touch_command(args: Dict[str, Any], client_func: Callable, action: str) -> CommandResults:
+    """Try setting caller as a reqistered caller.
+    If caller is not registered, set the caller argument as caller name.
+    This function implements incident_create and incident_update commands.
+    """
+    try:
+        args['registered_caller'] = True
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:616]: registered caller is a caller that has an account on the TOPdesk system and we can link its' account to the ticket. unregistered caller is someone without an account.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py (hunk 1 above)
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:697]: Screen Shot 2021-03-14 at 10 55 18

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py (hunk 2 above)
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:721]: I think it has the option info of the timezone as well but the client uses UTC. This is the format used:
Screen Shot 2021-03-14 at 11 01 04

I think dateparser will handle both

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py (hunk 3 above)
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:737]: This was actually copied from ServiceNow, I thought that the mapper uses these labels in the first place. I'll try it your way, see what happens.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 15]
```diff
@@ -0,0 +1,862 @@
...
+
+
+def attachment_upload_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    """Upload attachment to certain incident in TOPdesk"""
+    file_entry = args.get('file', None)
+    if not file_entry:
+        raise Exception("Found no file entry.")
+
+    file_name = demisto.dt(demisto.context(), "File(val.EntryID=='" + file_entry + "').Name")
+    if not file_name:  # in case of info file
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:645]: https://xsoar.pan.dev/docs/integrations/context-standards-mandatory#infofile

[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:645]: Could be a report just like the customer mentioned.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 16]
```diff
@@ -0,0 +1,862 @@
...
+            url_suffix=endpoint,
+            json_data=request_params
+        )
+
+    def incident_do(self, action: str,
+                    incident_id: Optional[str],
+                    incident_number: Optional[str],
+                    reason_id: Optional[str]) -> Dict[str, Any]:
+        """Preform action on TOPdesk incident with specified reason_id if needed."""
+        allowed_actions = ["escalate", "deescalate", "archive", "unarchive"]
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:131]: Could you elaborate? Not really used to using enums in cases like this

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md [hunk 17]
```diff
@@ -0,0 +1,4 @@
+## TOPdesk
+- Username and password should be the same as in the TOPdesk login.
+- Login with an operator account to manage incidents.
+- Make sure the account used has permissions to make all needed requests.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md:4]: I'm not really sure myself, I think the TOPdesk administrator handles this. I tried finding this information inside the TOPdesk system and on google but had no luck. Anyway I rewrote this file like I understood, hope its better now.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 18]
```diff
@@ -0,0 +1,862 @@
...
+    """main function, parses params and runs command functions."""
+
+    # get the service API url
+    demisto_params = demisto.params()
+    base_url = urljoin(demisto_params['url'], '/api')
+    verify_certificate = not demisto_params.get('insecure', False)
+
+    demisto.debug(f'Command being called is {demisto.command()}')
+    try:
+        encoded_credentials = b64encode(bytes(f"{demisto.params().get('username')}:{demisto.params().get('password')}",
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:784]: The API only uses basic auth which this implements. There is a python lib 'basicauth' that does that for you, the usage looks like this:
`encoded_credentials = basicauth.encode(username=demisto.params().get('username'),
                                               password=demisto.params().get('password'))

 headers = {'Authorization': encoded_credentials}`
 
 but our docker image doesn't have it :( 
 If you think the way it is is ugly enough to justify renewing the image let me know, I'll have it done.

[the remark this answers is not in the document]
Comment (Itay4) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:784]: @DinaMeylakh you should be able to pass a tuple of the username and password to the `requests` auth argument

[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py]: @Itay4 found it, thanks :)

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py (hunk 4 above)
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py]: I redid this so the millisecond and timezone will be taken into account

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py [hunk 19]
```diff
@@ -0,0 +1,816 @@
...
+        verify=False,
+        auth=('some_username', 'some_password'),
+        new_query=True
+    )
+    requests_mock.get(
+        command_api_request[0], json=[{}])
+    command(client, command_args)
+
+    assert requests_mock.called
+    assert requests_mock.last_request.json() == command_api_request[1]
```
Comment (Itay4) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:583]: not sure i understand why this is needed, these are all get requests with no body

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py [hunk 20]
```diff
@@ -0,0 +1,816 @@
...
+    action = "updating"
+    if create_func:
+        client_func = client.create_incident
+        request_method = "post"
+        action = "creating"
+    mock_topdesk_node = util_load_json(mock_response_file)
+    response_incident = mock_topdesk_node.copy()
+    request_command = getattr(requests_mock, request_method)
+
+    def callback_func(request, context):
```
Comment (Itay4) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:475]:
```suggestion
    def callback_func(request, _):
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py [hunk 21]
```diff
@@ -0,0 +1,816 @@
...
+        'creationDate': '2020-03-10T06:32:36Z',
+        'will_be_fetched': True
+    }], '2020-02-11T06:32:36.303000+0000', '2020-03-10T06:32:36.000000+0000'),
+    ([{  # Last fetch is at incident creation
+        'number': 'TEST-1',
+        'creationDate': '2020-02-10T06:32:36.303+0000',
+        'will_be_fetched': False
+    }], '2020-02-10T06:32:36.303000+0000', '2020-02-10T06:32:36.303000+0000'),
+])
+def test_fetch_incidents(requests_mock, topdesk_incidents_override, last_fetch_time, updated_fetch_time):
```
Comment (Itay4) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:767]: does it make sense to add test for 2 consecutive calls to `fetch_incidents` (could be same data returned) to ensure duplicates are not fetched?

**Review by Itay4 — COMMENTED** [review]
reviewed the unit tests - nice work on the high code coverage !
to be honest im not a big fan of the generalization you did here because its not easy to follow, so lets try at least to elaborate in the test docstring what we're testing in each case of the parameterize (where it makes sense to, for example `test_unsupported_old_query_param`)

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py (hunk 19 above)
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:583]: This is good for testing the assembly of the inline parameters of the request. Since the logic of this assembly is done within the integration code I felt the need to check it, especially since the query parameter is a bit tricky at times. If you feel this is redundant, I'll delete it.

Comment (Itay4) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_test.py:558]: you can keep it if you'd like, just wondered

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md [hunk 22]
```diff
@@ -0,0 +1,4 @@
+## TOPdesk
+- Username and password should be the same as in the TOPdesk login.
+- Login with an operator account to manage incidents.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md]: What about now?

**Review by roysagi — CHANGES_REQUESTED** [review]
Looks good!
Please see my notes and create a readme file - then we should assign @kirbles19 here to check the docs afterward.

**Comment by roysagi** [comment]
@DinaMeylakh - Also please add a test playbook here.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 23]
```diff
@@ -0,0 +1,5920 @@
...
+  required: false
+  type: 0
+- defaultvalue: 3 days
+  display: First fetch time (<number> <time unit>, e.g., 12 hours, 7 days, 3 months,
+    1 year)
+  name: first_fetch
+  required: false
+  type: 0
+- display: The query to use when fetching incidents
+  additionalinfo: Getting incidents with new style FIQL query is available only from TOPdeskRestAPI version 3.4.0. For earlier versions this field will be used as additional inline params as supported.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:40]: You know a way I can link it inside the YML? I tried to do it like in markdown but it didn't work, couldn't find anything useful online either. Will add links to the readme though.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md [hunk 24]
```diff
@@ -0,0 +1,23 @@
+## TOPdesk
+
+Some of TOPdesk's user types include Customer and Operator. Each user type has a different set of actions that he is allowed to perfom.
+For example, a customer can only update incidents he created while an operator can update a wider range of incidents.
+
+In general, if an account can preform the needed command from TOPdesk it should be able to preform it with this integration. 
+- Make sure you use the right account for your needs and that the account used has permissions to make all needed requests.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md:7]: Added a section to each command in the README.md and a refrence to there here.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml [hunk 25]
```diff
@@ -0,0 +1,5920 @@
...
+  name: password
+  required: true
+  type: 4
+- display: Fetch incidents
+  name: isFetch
+  required: false
+  type: 8
+- defaultvalue: true
+  display: Use new style FIQL query when listing branches, persons and operators. (e.g. query=status==firsLine)
+  additionalinfo: Getting lists with new style FIQL query is available from SupportingFilesAPI version 1.38.0. If the checkbox is unchecked, old style queries will be used. Filtering TOPdesk incidents will not be affected by this configuration.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.yml:24]: The information about the SupportingFilesAPI version is not available anywhere to my knowledge. Tried getting it from the UI but it's not there. TOPdesk team hasn't replied to my email about this either.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 26]
```diff
@@ -0,0 +1,1165 @@
...
+
+    latest_created_time = last_fetch_datetime
+    incidents: List[Dict[str, Any]] = []
+
+    modification_date_start = last_fetch_datetime.strftime("<%Y-%m-%d>")
+
+    topdesk_incidents = get_incidents_with_pagination(client=client,
+                                                      max_fetch=int(demisto_params.get('max_fetch', 10)),
+                                                      query=demisto_params.get('query', None),
+                                                      modification_date_start=modification_date_start)
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:1026]: Wouldn't the modified incidents just be updated in XSOAR? They should all still have the same id's
There is an analogous creation_date_start and creation_date_end if needed

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 27]
```diff
@@ -0,0 +1,1165 @@
...
+        if int(last_fetch_datetime.timestamp()) < int(incident_created_time.timestamp()):
+
+            incident = {
+                'name': f"TOPdesk incident {topdesk_incident['number']}",
+                'details': json.dumps(topdesk_incident),
+                'occurred': incident_created_time.strftime(DATE_FORMAT),
+                'rawJSON': json.dumps(topdesk_incident)}
+
+            if incident not in incidents:  # Do not fetch duplicates
+                incidents.append(incident)
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:1043]: dictionaries are unhashable and therefore cannot be added to sets :( but I liked your idea

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md [hunk 28]
```diff
@@ -0,0 +1,23 @@
+## TOPdesk
+
+Some of TOPdesk's user types include Customer and Operator. Each user type has a different set of actions that he is allowed to perfom.
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md:3]: Added a section to each command in the README.md. They do, it differs for each command.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 29]
```diff
@@ -0,0 +1,1165 @@
...
+    else:
+        last_fetch_datetime = dateparser.parse(last_fetch)
+
+    latest_created_time = last_fetch_datetime
+    incidents: List[Dict[str, Any]] = []
+
+    modification_date_start = last_fetch_datetime.strftime("<%Y-%m-%d>")
+
+    topdesk_incidents = get_incidents_with_pagination(client=client,
+                                                      max_fetch=int(demisto_params.get('max_fetch', 10)),
```
[the remark this answers is not in the document]
Comment (DinaMeylakh) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:1024]: ServiceNow also has a default of 10, also this is the default for TOPdesk API get-incidents command

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:1023]: If I am getting your question correctly you are talking about the mirroring feature. This is a separate process, when fetching we need to aim to bring each unique incident once - after that, if mirroring is enabled (and of course set up correctly in the code) then there is a flow for mirror in and out which happens for every incident on its own.

or simply - fetch incidents is not updating existing incidents, only brings new ones into the system.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md:8]: Still, this needs to appear here - at least a comment about checking the docs about the list of the required permissions for every command.

File: Packs/TOPdesk/Integrations/TOPdesk/README.md [hunk 30]
```diff
@@ -0,0 +1,3707 @@
+TOPdesk’s Enterprise Service Management software (ESM) lets your service teams join forces and process requests from a single platform.
+This integration was integrated and tested with 10.08.008-on-premises-release3-build2 for Linux x86 64-bit of TOPdesk
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/README.md:2]:
```suggestion
This integration was integrated and tested with 10.08.008-on-premises-release3-build2 for Linux x86 64-bit of TOPdesk.
```

File: Packs/TOPdesk/Integrations/TOPdesk/README.md [hunk 31]
```diff
@@ -0,0 +1,3707 @@
...
+
+#### Troubleshooting
+- Make sure the application password is not expired by logging in TOPdesk and viewing it as described in the **Setup TOPdesk's application password** section step 3. 
+
+### FIQL query
+A few implemented commands can take a query as a paramter. Those include: 
+`topdesk-incidents-list`, `topdesk-branches-list`, `topdesk-persons-list`, `topdesk-operators-list`
+
+While the newest versions of TOPdesk all use the new [FIQL query format](https://developers.topdesk.com/tutorial.html#query), older versions use limited inline parameters for filters in the requests.
+More specifically there are 2 versions at work:
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/README.md:59]:
```suggestion
Specifically there are 2 versions at work:
```

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md [hunk 32]
```diff
@@ -0,0 +1,25 @@
+## TOPdesk
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk_description.md:1]: We should note here that this integration supports the API from version 1.38.0 and to see the documentation for additional information.

File: Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py [hunk 33]
```diff
@@ -0,0 +1,1303 @@
...
+            else:
+                new_query_args.append('='.join(query_arg_list))
+
+        return '&'.join(new_query_args)
+
+
+''' HELPER FUNCTIONS '''
+
+
+def trim_results_by_limit(results: List[Any], limit: Union[int, str] = 100) -> List[Any]:
```
Comment (roysagi) [inline Packs/TOPdesk/Integrations/TOPdesk/TOPdesk.py:377]: the issue with this command is that you still bring all the info locally so you don't save the effort for the large HTTP request - is there no `limit` option from the API itself? (if not then this is good and you can resolve this comment.)

**Review by roysagi — APPROVED** [review]
Looks good @DinaMeylakh!
Great job!
See my minor notes.

**Comment by Michal-Vardi** [comment]
@DinaMeylakh 
Done
